cpp/src/round_robin/round_robin.cu

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright (c) 2019-2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <thrust/execution_policy.h>
+#include <thrust/for_each.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/transform_iterator.h>
+#include <thrust/tuple.h>
+#include <thrust/device_vector.h>
+#include <thrust/scan.h>
+
+#include <algorithm>
+#include <utility>
+#include <vector>
+#include <memory>
+#include <type_traits>
+#include <cmath> // for std::ceil()
+
+#include <cudf/types.hpp>
+#include <cudf/table/table.hpp>
+#include <cudf/table/table_device_view.cuh>
+#include <cudf/table/row_operators.cuh>
+#include <cudf/utilities/type_dispatcher.hpp>
+#include <rmm/thrust_rmm_allocator.h>
+#include <cudf/utilities/bit.hpp>
+#include <cudf/null_mask.hpp>
+#include <cudf/copying.hpp>
+#include <cudf/detail/utilities/cuda.cuh>
+#include <cudf/detail/gather.cuh>
+
+namespace cudf {
+namespace experimental { 
+namespace detail {
+
+std::pair<std::unique_ptr<table>,
+          std::vector<cudf::size_type>>
+round_robin_partition(table_view const& input,
+                      cudf::size_type num_partitions,
+                      cudf::size_type start_partition = 0,
+                      rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource(),
+                      cudaStream_t stream = 0)
+{
+  auto nrows = input.num_rows();
+
+  CUDF_EXPECTS( num_partitions > 1 && num_partitions < nrows, "Incorrect number of partitions. Must be greater than 1 and less than number of rows." );
+  CUDF_EXPECTS( start_partition < num_partitions, "Incorrect start_partition index. Must be less than number of partitions." );
+  
+  auto n_pmax = nrows % num_partitions;//# partitions of max size
+  cudf::size_type max_p_size = std::ceil( static_cast<double>(nrows) / static_cast<double>(num_partitions));// max size of partitions
+  
+  auto pmm = n_pmax * max_p_size;
```

Comment:
These variables should be given descriptive, human readable names.

Reply:
Done.